system/settings/settings.c

```diff
@@ -0,0 +1,1717 @@
+  g_settings.sev.sigev_value.sival_ptr = &g_settings.wrpend;
+  g_settings.sev.sigev_notify_function = dump_cache;
+
+  memset(&g_settings.trigger, 0, sizeof(struct itimerspec));
+  g_settings.trigger.it_value.tv_sec   =
+                               CONFIG_SYSTEM_SETTINGS_CACHE_TIME_MS / 1000;
+  g_settings.trigger.it_value.tv_nsec  =
+                              (CONFIG_SYSTEM_SETTINGS_CACHE_TIME_MS % 1000) *
+                               1000 * 1000;
+
+  timer_create(CLOCK_REALTIME, &g_settings.sev, &g_settings.timerid);
+#endif
+  g_settings.initialized = true;
+  g_settings.hash = 0;
+  g_settings.wrpend = false;
+
+  return OK;
+}
+
+/****************************************************************************
+ * Name: settings_setstorage
+ *
+ * Description:
+ *    Sets a file to be used as a settings storage.
+ *    Except from the first file, if loading the file causes any changes
+ *    to the settings, then the new map will be dumped to all files
+ *    (effectively it syncs all storages).
+ *
+ * Input Parameters:
+ *    file             - the filename of the storage to use
+ *    type             - the type of the storage (BINARY or TEXT)
+ *
+ * Returned Value:
+ *   Success or negated failure code
+ *
+ ****************************************************************************/
+
+int settings_setstorage(FAR char *file, enum storage_type_e type)
+{
+  FAR storage_t *storage = NULL;
+  int ret = OK;
+  int idx = 0;
+  uint32_t h;
+
+  ret = pthread_mutex_lock(&g_settings.mtx);
+  if (ret < 0)
+    {
+      return ret;
+    }
+
+  if (!g_settings.initialized)
+    {
+      ret = -EACCES;
+      goto errout;
+    }
+
+  while ((idx < CONFIG_SYSTEM_SETTINGS_MAX_STORAGES) &&
+         (g_settings.store[idx].file[0] != '\0'))
+    {
+      idx++;
+    }
+
+  DEBUGASSERT(idx < CONFIG_SYSTEM_SETTINGS_MAX_STORAGES);
+  if (idx >= CONFIG_SYSTEM_SETTINGS_MAX_STORAGES)
+    {
+      ret = -ENOSPC;
+      goto errout;
+    }
+
+  DEBUGASSERT(strlen(file) < CONFIG_SYSTEM_SETTINGS_MAX_FILENAME);
+  if (strlen(file) >= CONFIG_SYSTEM_SETTINGS_MAX_FILENAME)
+    {
+      ret = -EINVAL;
+      goto errout;
+    }
+
+  storage = &g_settings.store[idx];
+  strncpy(storage->file, file, sizeof(storage->file));
+  storage->file[sizeof(storage->file) - 1] = '\0';
+
+  switch (type)
+  {
+    case STORAGE_BINARY:
+      {
+        storage->load_fn = load_bin;
+        storage->save_fn = save_bin;
+        storage->size_fn = size_bin;
+      }
+      break;
+
+    case STORAGE_TEXT:
+      {
+        storage->load_fn = load_text;
+        storage->save_fn = save_text;
+        storage->size_fn = size_text;
+      }
+      break;
+
+    case STORAGE_EEPROM:
+      {
+        storage->load_fn = load_eeprom;
+        storage->save_fn = save_eeprom;
+        storage->size_fn = size_eeprom;
+      }
+      break;
+
+    default:
+      {
+        DEBUGASSERT(0);
+      }
+      break;
+  }
+
+  ret = storage->load_fn(storage->file);
```

Comment:
This should return an error.
It is perfectly OK to set-up a new storage that just happens to be unavailable for the moment.

For example, during initialization I set-up a file in the SD card. But the card may not be present at the time!
The storage is properly configured though, and it will be used at a later time, when the card is inserted.

Reply:
the load might fail due to invalid crc or "valid" though? The return value indicates the reason for the "fail" and it is up to the user app to deal with it. It is not preventing the storage being set up, is it?

Reply:
I guess this should be handled by syncing of the storages?

Let me propose something.
What if `load()`, `save()`, `sync()` etc return OK if *at least* one storage succeeds, and error if *all* of them fail?

Reply:
I think this might be a little simpler?

- calls to the save functions are via dump_cache, which is a void function as it can be called on a timer.
  - So the save can be a void function as we'll never know about it.
- Calls to the load function are only made from settings_sync or settings_setstorage 
  -  settings_setstorage is called for a single storage so that would be the time to deal with a failed load.
      - but this should "bail" if the (single storage file) load fails so we do pick up the failed load 
  - settings_sync does load both storages so it is fair that we only return a fail if BOTH storages fail I think. Any fail of a single store will be picked up on the next settings_setstorage, probably the next time the app restarts?


```
ret = settings_setstorage(our_storage, our_storage_type);
if (ret < 0)
  {
    /* look at fail reason and do something if we're at all bothered about this fail */
  
ret = settings_sync();
if (ret < 0)
  {
    /* all storages failed to load - DO SOMETHING  if we are at all bothered */
  }
```

Does that make sense? I've had 3 different thoughts on this now so I might have lost my way!!

Reply:
I have implemented the above as I think it makes sense.